pkg/opensearch/opensearch.go

```diff
@@ -64,11 +65,53 @@ func (ds *OpenSearchDatasource) QueryData(ctx context.Context, req *backend.Quer
 		return nil, err
 	}
 
+	err = handleServiceMapPrefetch(ctx, osClient, req)
+	if err != nil {
+		return nil, err
+	}
+
 	query := newQueryRequest(osClient, req.Queries, req.PluginContext.DataSourceInstanceSettings, intervalCalculator)
 	response, err := wrapError(query.execute(ctx))
 	return response, err
 }
 
+// handleServiceMapPrefetch inspects the given request, and, if it wants a serviceMap, creates and
+// calls the Prefetch query to get the services and operations lists that are required for
+// the associated Stats query. It then adds these parameters to the originating query so
+// the Stats query can be created later.
+func handleServiceMapPrefetch(ctx context.Context, osClient client.Client, req *backend.QueryDataRequest) error {
+	var serviceMapQuery backend.DataQuery
+	var serviceMapQueryIndex int
+	for i, query := range req.Queries {
+		model, _ := simplejson.NewJson(query.JSON)
+		queryType := model.Get("queryType").MustString()
+		luceneQueryType := model.Get("luceneQueryType").MustString()
+		serviceMapRequested := model.Get("serviceMap").MustBool(false)
+		if queryType == Lucene && luceneQueryType == "Traces" && serviceMapRequested {
+			serviceMapQueryIndex = i
+			serviceMapQuery = createServiceMapPrefetchQuery(query)
+			break
+		}
+	}
+	if serviceMapQuery.JSON != nil {
+		query := newQueryRequest(osClient, []backend.DataQuery{serviceMapQuery}, req.PluginContext.DataSourceInstanceSettings, intervalCalculator)
+		response, _ := wrapError(query.execute(ctx))
+		services, operations := extractParametersFromServiceMapFrames(response)
+
+		// encode the services and operations back to the JSON of the query to be used in the stats request
+		model, _ := simplejson.NewJson(req.Queries[serviceMapQueryIndex].JSON)
```

Comment:
should we use `_` here? Seems like we could conceivably have an error here and deal with it?

Reply:
I don't think we expect an error here, since it's a non-modified request coming from FE, but it can't hurt to add it!

Reply:
In this case it is impossible for there to be an error - we've already called `simplejson.NewJson` on the `JSON` for every query, including this one, on line 86. I like leaving the `_` here to indicate that, but I can add a comment as well.

Reply:
(Well - impossible barring concurrency shenanigans, which I don't _think_ can happen here)